lean4/src/putnam_2017_a2.lean

```diff
@@ -1,11 +1,14 @@
 import Mathlib
 
+open scoped Polynomial
+
 /--
 Let $Q_0(x)=1$, $Q_1(x)=x$, and $Q_n(x)=\frac{(Q_{n-1}(x))^2-1}{Q_{n-2}(x)}$ for all $n \geq 2$. Show that, whenever $n$ is a positive integer, $Q_n(x)$ is equal to a polynomial with integer coefficients.
 -/
 theorem putnam_2017_a2
-(Q : ℕ → ℝ → ℝ)
-(hQbase : ∀ x : ℝ, Q 0 x = 1 ∧ Q 1 x = x)
-(hQn : ∀ n ≥ 2, ∀ x : ℝ, Q n x = ((Q (n - 1) x) ^ 2 - 1) / Q (n - 2) x)
-: ∀ n > 0, ∃ P : Polynomial ℝ, (∀ i : ℕ, P.coeff i = round (P.coeff i)) ∧ Q n = P.eval :=
-sorry
+    (Q : ℕ → ℝ → ℝ)
+    (hQbase : ∀ x : ℝ, Q 0 x = 1 ∧ Q 1 x = x)
+    (hQn : ∀ n, ∀ x, Q n x ≠ 0 → Q (n + 2) x = ((Q (n + 1) x) ^ 2 - 1) / Q n x)
```

Comment:
Nice catch. I am not sure the statement is true still, though. Does this not allow the values of Q (n+2) x be essentially unconstrained when Q n x = 0? Maybe this is better stated as operations of formal polynomials?

Reply:
Good point! I've restated it using rational functions to fix this.